src/shared/tpm2-util.c

```diff
@@ -307,55 +318,232 @@ static int tpm2_credit_random(ESYS_CONTEXT *c) {
+
+        assert(ret_primary);
+
+        TSS2_RC rc = sym_Esys_GetCapability(c,
+                        ESYS_TR_NONE,
+                        ESYS_TR_NONE,
+                        ESYS_TR_NONE,
+                        TPM2_CAP_HANDLES,
+                        SRK_HANDLE,
+                        1,
+                        &more_data,
+                        &cap_data);
+        if (rc != TSS2_RC_SUCCESS)
+                return log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
+                                       "Failed to enumerate handles searching for SRK: %s",
+                                       sym_Tss2_RC_Decode(rc));
+
+        /* Did Not find SRK, indicate this by returning 0 and ESYS_TR_NONE */
+        if (cap_data->data.handles.count == 0) {
+                *ret_primary = ESYS_TR_NONE;
+
+                if (ret_alg)
+                        *ret_alg = 0;
+                return 0;
+        }
+
+        log_debug("Found SRK on TPM.");
+
+        /* convert the raw handle to an ESYS_TR */
+        TPM2_HANDLE handle = cap_data->data.handles.handle[0];
+        rc = sym_Esys_TR_FromTPMPublic(c,
+                        handle,
+                        ESYS_TR_NONE,
+                        ESYS_TR_NONE,
+                        ESYS_TR_NONE,
+                        &primary);
+        if (rc != TSS2_RC_SUCCESS)
+                return log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
+                                        "Failed to convert ray handle to ESYS_TR for SRK: %s",
+                                        sym_Tss2_RC_Decode(rc));
+
+        /* Make sure the alg is what we were expecting */
+        _cleanup_(Esys_Freep) TPM2B_PUBLIC *out_public = NULL;
+        _cleanup_(Esys_Freep) TPM2B_NAME *out_name = NULL, *out_qname = NULL;
+        rc = sym_Esys_ReadPublic(
+                        c,
+                        primary,
+                        ESYS_TR_NONE,
+                        ESYS_TR_NONE,
+                        ESYS_TR_NONE,
+                        &out_public,
+                        &out_name,
+                        &out_qname);
+        if (rc != TSS2_RC_SUCCESS)
+                return log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
+                                        "Failed to convert ray handle to ESYS_TR for SRK: %s",
+                                        sym_Tss2_RC_Decode(rc));
+
+        /* Cannot specify the alg and have a mismatch */
+        if (alg != 0 && alg != out_public->publicArea.type)
+                return log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
+                                        "SRK expected alg 0x%x does not match found alg 0x%x",
+                                        alg, out_public->publicArea.type);
+
+        *ret_primary = primary;
+
+        if (ret_alg)
+                 *ret_alg = out_public->publicArea.type;
+
+        return 0;
+}
+
+
+static int tpm2_make_transient_primary(
```

Comment:
it seems wrong to name this 'make transient primary' when it's actually persisting an SRK...

Reply:
> it seems wrong to name this 'make transient primary' when it's actually persisting an SRK...

That was @poettering 's suggestion so I didn't fight it.

Reply:
I changed it back